### PR TITLE
Improve addLastJoin behaviour

### LINKED ISSUE #1737: SQLQuery.toString should not have side effects
Calling ProjectableSQLQuery.toString registers pending join clauses as a side effect. If one happens to call query.toString() between join() and on() then the query breaks as a result: 

    query.from(x);
    query.leftJoin(y) // toString() call e.g. by debugger
    .on(x.foo.equals(y.foo));
    query.leftJoin(z).on(z.foo.equals(y.foo));

Results in 

    from x 
    left join y 
    left join z on x.foo = y.foo and z.foo = y.foo

ToString with side effects is especially disastrous with modern debuggers as they call toString every time execution pauses. This makes debugging query builders very hard and error-prone.

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1737

### CODE REVIEW

File: querydsl-core/src/test/java/com/querydsl/core/DefaultQueryMetadataTest.java [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 import java.util.Arrays;
 
+import com.google.common.collect.ImmutableSet;
```
Comment (Shredder121) [inline querydsl-core/src/test/java/com/querydsl/core/DefaultQueryMetadataTest.java]: Did you change your import settings?

Comment (timowest) [inline querydsl-core/src/test/java/com/querydsl/core/DefaultQueryMetadataTest.java]: Sorry, I forgot to configure them on my new machine

Comment (Shredder121) [inline querydsl-core/src/test/java/com/querydsl/core/DefaultQueryMetadataTest.java]: Ah, that explains it.
I was just wondering.

**Comment by Shredder121** [comment]
Looks good now.

**Comment by robertandrewbain** [comment]
Looks good.

**Pull request merged**